source/common/common/status.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include <atomic>
+#include <string>
+
+#include "envoy/http/codes.h"
+
+#include "absl/status/status.h"
+#include "absl/strings/string_view.h"
+
+/**
+ * Facility for returning rich error information.
+ * This facility is to be used in place of exceptions, in components where
+ * exceptions safety is not guaranteed (i.e. codecs).
+ *
+ * Envoy::Status is an alias of absl::Status.
+ * IMPORTANT: `absl::Status` constructor `absl::Status::code()` and absl::Status::ToString()`
+ * methods must not be used as they will not return correct error information. Instead the error
+ * value creating and corresponding error checking functions defined below must be used.
+ * TODO(yanavlasov): add clang-tidy or lint check to enforce this.
+ *
+ * Usage example:
+ *
+ *  Envoy::Status Foo() {
+ *    ...
+ *    if (codec_error) {
+ *      return CodecProtocolError("Invalid protocol");
+ *    }
+ *    return Envoy::OkStatus();
+ *  }
+ *
+ *  void Bar() {
+ *    auto status = Foo();
+ *    if (status.ok()) {
+ *      ...
+ *    } else {
+ *      ASSERT(IsCodecProtocolError(status));
+ *      ENVOY_LOG(debug, "Codec error encountered: {}", status.message());
+ *    }
+ *  }
+ */
+
+namespace Envoy {
+
+/**
+ * Status codes for representing classes of Envoy errors.
```

Comment:
Can you add a reference to source/common/http/exception.h for more info (or copy comments about what each error represents?)

Reply:
done